### PR TITLE
Build special win-x86 JIT with funclets for NativeAOT

### PR DESCRIPTION
**Pull request by filipnavara**

### CODE REVIEW

**Comment by filipnavara** [comment]
Not sure what I messed up. I'll run it again, locally...

File: src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs [hunk 1]
```diff
@@ -142,10 +141,19 @@ private static string GetTargetSpec(TargetDetails target)
                 _ => throw new NotImplementedException(target.Architecture.ToString())
             };
 
-            if ((target.Architecture == TargetArchitecture.ARM64) || (target.Architecture == TargetArchitecture.ARM))
+            string targetOSComponent;
+            if (target.Architecture is TargetArchitecture.ARM64 or TargetArchitecture.ARM)
             {
                 targetOSComponent = "universal";
             }
+            else if (target.OperatingSystem == TargetOS.Windows && target.Architecture == TargetArchitecture.X86 && target.Abi == TargetAbi.NativeAot)
```
Comment (filipnavara) [inline src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs:149]: Hmm. R2R sets the ABI to `Crossgen2RootCommand.IsArmel ? TargetAbi.NativeAotArmel : TargetAbi.NativeAot`. I did not expect that.

Comment (filipnavara) [reply src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs:149]: @MichalStrehovsky is this intentional?

As far as I can see the expected value would be `TargetAbi.Jit` (ie. match what `CorInfoImpl` reports to RyuJIT). There are couple of checks for unsupported ARMel platform and this one branch, which apparently doesn't take place today:
https://github.com/dotnet/runtime/blob/b29efed9995da044d97f9d3f97bc3d9de80a2050/src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs#L954-L955

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs:149]: TargetAbi.Jit is for the native AOT JIT prototype. Dead code. The ABI=armel is also weird but sigh.

Ref: https://github.com/dotnet/runtime/pull/58552#discussion_r701584141
Ref: https://github.com/dotnet/runtime/pull/43706#pullrequestreview-515494769

You can use the `#if READYTORUN` ifdef to decide between crossgen2 and ilc in this file.

**Comment by filipnavara** [comment]
cc @dotnet/jit-contrib

**Comment by BruceForstall** [comment]
Is this part of a larger effort to support win-x86 on NativeAOT? Is there a design spec (issue) somewhere? @MichalStrehovsky

**Comment by filipnavara** [comment]
> Is this part of a larger effort to support win-x86 on NativeAOT?

Yes. I have private branch which compiles and passes the smoke tests. I'm in process of upstreaming it piece by piece.

**Comment by filipnavara** [comment]
> Is there a design spec (issue) somewhere?

Tracking issue: https://github.com/dotnet/runtime/issues/86573

The prototype uses x86 custom GC info, EH funclets (hence this PR), and Fastcall-like ABI.

**Comment by MichalStrehovsky** [comment]
> Is this part of a larger effort to support win-x86 on NativeAOT? Is there a design spec (issue) somewhere? @MichalStrehovsky

This is a community port that restores the Windows x86 support inherited from .NET Native.

.NET Native on x86 uses funclets, so the entire runtime assumes it; if we can get codegen to do it, we don't need to rewrite exception handling in the runtime.

**Comment by filipnavara** [comment]
FWIW I started with the SEH first but it quickly turned into a mess since part of the exception handling is in native code and part in the managed code. There was little potential for code reuse between CoreCLR and NativeAOT.

The funclet approach has multiple benefits. Nearly all the exception handling code in NativeAOT readily works. The code can be shared between Windows and Linux if we decide to port to Linux later. The VM unwinding code can be shared almost entirely between CoreCLR and NativeAOT and it already implements the funclet code paths thanks to linux-x86 port.

**Review by BruceForstall — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Is this new JIT included as part of the crossgen2 distribution? We should ensure that it isn't (@davidwrighton spent a bunch of effort to minimize the number of JITs built and distributed).

**Comment by filipnavara** [comment]
> Is this new JIT included as part of the crossgen2 distribution?

It's not supposed to be. I'm not sure how to easily check.

**Pull request merged**